### PR TITLE
Handle UISceneDelegate changes in Auth

### PR DESCRIPTION
**Pull request by renkelvin**

### CODE REVIEW

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m [hunk 1]
```diff
@@ -1002,6 +1064,47 @@ + (void)proxyAppDelegate:(id<GULApplicationDelegate>)appDelegate {
   }
 }
 
++ (void)proxySceneDelegateWithClass:(Class)sceneDelegateClass API_AVAILABLE(ios(13.0)) {
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
```
Comment (maksymmalyhin) [inline GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: It looks like `dispatch_once` will allow only the first scene delegate to be swizzled, so other will be ignored.

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m [hunk 2]
```diff
@@ -1002,6 +1064,47 @@ + (void)proxyAppDelegate:(id<GULApplicationDelegate>)appDelegate {
...
+    IMP orignalImplementation = method_getImplementation(originalMethod);
+    NSValue *originalImplementationPointer = [NSValue valueWithPointer:orignalImplementation];
+    NSDictionary *sceneDelegateSELIMPPair = @{NSStringFromSelector(methodSelector): originalImplementationPointer};
+    sceneDelegateIMPDict[sceneDelegateClassName] = sceneDelegateSELIMPPair;
+
+    objc_setAssociatedObject(appDelegate, &kGULSceneDelegateIMPDictKey,
+                             [sceneDelegateIMPDict copy], OBJC_ASSOCIATION_RETAIN);
+
+    BOOL didAddMethod =
+    class_addMethod(sceneDelegateClass,
```
Comment (maksymmalyhin) [inline GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: As far as I can see, the method swizzling has been implemented here. ISA swizzling is used to proxy App Delegate which is a preferable way (method swizzling will affect any object of the class in contrast to ISA swizzling which will affect only a particular instance, so the impact on the host app will be smaller as well as risks of issues). 

I think, [GULObjectSwizzler](https://github.com/firebase/firebase-ios-sdk/blob/master/GoogleUtilities/ISASwizzler/Private/GULObjectSwizzler.h) can be used here to create a subclass for the object. Then [proxyDestinationSelector...](https://github.com/firebase/firebase-ios-sdk/blob/e88b52fb676d86df0bf665f27a899bfed46efc9c/GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m#L575) method may be used to update the method implementation.

Feel free to reach out to me to clarify the details.

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m (hunk 1 above)
Comment (renkelvin) [reply GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: Yeah, good catch!

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m (hunk 2 above)
Comment (renkelvin) [reply GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: Please see the updated pr. Seems works well. I'll add unit tests later.

Comment (maksymmalyhin) [reply GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: Yeah, I also would feel more comfortable with the changes having tests in place.

Comment (renkelvin) [reply GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: I mean this pr is more of a draft. The solution is not finalized yet.

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m [hunk 3]
```diff
@@ -1002,6 +1060,85 @@ + (void)proxyAppDelegate:(id<GULApplicationDelegate>)appDelegate {
   }
 }
 
++ (void)proxySceneDelegate:(UIScene *)scene API_AVAILABLE(ios(13.0)) {
+  Class realClass = [scene.delegate class];
+
+  NSString *classNameWithPrefix =
+  [kGULAppDelegatePrefix stringByAppendingString:NSStringFromClass(realClass)];
```
[the remark this answers is not in the document]
Comment (renkelvin) [reply GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: Nope, it's the built-in format

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m (hunk 2 above)
Comment (maksymmalyhin) [reply GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: Ah, sure, no problem

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m [hunk 4]
```diff
@@ -1002,6 +1060,85 @@ + (void)proxyAppDelegate:(id<GULApplicationDelegate>)appDelegate {
...
+  // cannot have more ivars/properties than its superclass since it will cause an offset in memory
+  // that can lead to overwriting the isa of an object in the next frame.
+  if (class_getInstanceSize(realClass) != class_getInstanceSize(sceneDelegateSubClass)) {
+    GULLogError(kGULLoggerSwizzler, NO,
+                [NSString stringWithFormat:@"I-SWZ%06ld",
+                 (long)kGULSwizzlerMessageCodeAppDelegateSwizzling007],
+                @"Cannot create subclass of App Delegate, because the created subclass is not the "
+                @"same size. %@",
+                NSStringFromClass(realClass));
+    NSAssert(NO, @"Classes must be the same size to swizzle isa");
```
[the remark this answers is not in the document]
Comment (renkelvin) [reply GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: @maksymmalyhin Asserting seems used a lot for AppDelegate, what do you think?

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m [hunk 5]
```diff
@@ -1002,6 +1060,85 @@ + (void)proxyAppDelegate:(id<GULApplicationDelegate>)appDelegate {
...
+  if (object_setClass(scene.delegate, sceneDelegateSubClass)) {
+    GULLogDebug(kGULLoggerSwizzler, NO,
+                [NSString stringWithFormat:@"I-SWZ%06ld",
+                 (long)kGULSwizzlerMessageCodeAppDelegateSwizzling008],
+                @"Successfully created App Delegate Proxy automatically. To disable the "
+                @"proxy, set the flag %@ to NO (Boolean) in the Info.plist",
+                [GULAppDelegateSwizzler correctAppDelegateProxyKey]);
+  }
+
+  // TODO: reassign delegate as app delegate above?
```
[the remark this answers is not in the document]
Comment (renkelvin) [reply GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: Yes, we need discuss more to see if it's needed.
https://github.com/firebase/firebase-ios-sdk/blob/2ebb82d1c511551748ebac4149e0dcdd090491af/GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m#L545

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m [hunk 6]
```diff
@@ -1002,6 +1060,85 @@ + (void)proxyAppDelegate:(id<GULApplicationDelegate>)appDelegate {
...
+                 (long)kGULSwizzlerMessageCodeAppDelegateSwizzling005],
+                @"Cannot create a proxy for App Delegate. Subclass already exists. Original Class: "
+                @"%@, subclass: %@",
+                NSStringFromClass(realClass), newClassName);
+    return;
+  }
+
+  // Register the new class as subclass of the real one. Do not allocate more than the real class
+  // size.
+  Class sceneDelegateSubClass = objc_allocateClassPair(realClass, newClassName.UTF8String, 0);
```
[the remark this answers is not in the document]
Comment (renkelvin) [reply GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: Yeah, it's a good point. If developer's app uses a prefix of "GUL_" like us, then there would be collision issue any way.

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m (hunk 5 above)
[the remark this answers is not in the document]
Comment (renkelvin) [reply GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: Removed it for now. I didn't see any issue without reassigning.

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m [hunk 7]
```diff
@@ -289,6 +296,23 @@ + (void)proxyOriginalDelegate {
     id<GULApplicationDelegate> originalDelegate =
         [GULAppDelegateSwizzler sharedApplication].delegate;
     [GULAppDelegateSwizzler proxyAppDelegate:originalDelegate];
+
+#if __IPHONE_OS_VERSION_MAX_ALLOWED >= 130000
```
[the remark this answers is not in the document]
Comment (renkelvin) [reply GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: What's the current status of this?

**Pull request closed without merging**